text/0000-dockerfiles.md

```diff
@@ -0,0 +1,113 @@
+
+Both Dockerfiles must accept `base_image` and `build_id` args. The `base_image` arg allows the lifecycle to specify the original base image. The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`.  
+
+A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of app-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before `run.Dockerfile` is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.
+
+### Platform-specified Dockerfiles
+
+The same Dockerfiles may be used to create new base images or modify existing base images outside of the app build process. For both app-specified and base-image-modifying Dockerfiles, any specified labels override existing values.
+
+Dockerfiles that are used to create a base image must create a `/cnb/image/genpkgs` executable that outputs a [CycloneDX](https://cyclonedx.org)-formatted list of packages in the image with PURL IDs when invoked. This executable is executed after any run.Dockerfile or build.Dockerfile is applied, and the output replaces the label `io.buildpacks.sbom`. This label doubles as a Software Bill-of-Materials for the base image. In the future, this label will serve as a starting point for the application SBoM.
+
+### Examples
+
+run.Dockerfile used to create a runtime base image:
+
+```
+ARG base_image
+FROM ${base_image}
+ARG build_id=0
+
+LABEL io.buildpacks.image.distro=ubuntu
+LABEL io.buildpacks.image.version=18.04
+LABEL io.buildpacks.rebasable=true
+
+ENV CNB_USER_ID=1234
+ENV CNB_GROUP_ID=1235
+
+RUN groupadd cnb --gid ${CNB_GROUP_ID} && \
+  useradd --uid ${CNB_USER_ID} --gid ${CNB_GROUP_ID} -m -s /bin/bash cnb
+
+USER ${CNB_USER_ID}:${CNB_GROUP_ID}
+
+COPY genpkgs /cnb/image/genpkgs
+```
+
+run.Dockerfile present in an app directory that always installs the latest version of curl:
+```
+ARG base_image
+FROM ${base_image}
+ARG build_id=0
+
+LABEL io.buildpacks.rebasable=true
+
+RUN echo ${build_id}
+
+RUN apt-get update && apt-get install -y curl && rm -rf /var/lib/apt/lists/*
+```
+
+Unsafe run.Dockerfile present in an app directory:
+```
+ARG base_image
+FROM ${base_image}
+ARG build_id=0
+
+LABEL io.buildpacks.rebasable=false
+
+RUN curl -L https://example.com/mypkg-install | sh # installs to /opt
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Involves breaking changes.
+- Buildpacks cannot install OS packages directly, only select runtime base images.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Stackpacks
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
```

Comment:
Will it be possible to execute a `run.Dockerfile` or `build.Dockerfile` with a `root` user to install additional packages (.eg. maven, jdk, ..) as such option was included within the RFC69 which is now closed ?

Reply:
My understanding from the working group discussion yesterday is that the dockerfiles execute as root, they can basically do anything to the container.

Reply:
> My understanding from the working group discussion yesterday is that the dockerfiles execute as root, they can basically do anything to the container.
and

> Only the Dockerfiles can run as root. This proposal does not allow buildpacks to run as root.

Can you elaborate please. Which application, system will then execute the command to build as root the Dockerfiles ?

Reply:
I don't recall that being discussed in the meeting. I would have to defer to @sclevine

Reply:
The platform would likely orchestrate new lifecycle phases that apply the Dockerfiles. This RFC does not (yet) propose an implementation, just UX.

Reply:
typically the instrucitons in a dockerfile would run as the `USER` set on the base image. Would we expect Dockerfiles to explicilty set `USER root` and then switch back to `USER cnb`?

Reply:
As proposed, yes, but I could definitely be convinced otherwise. We could restore the original UID/GID if the final values are 0:0.